src/main/java/bio/terra/janitor/service/cleanup/FlightManager.java

```diff
@@ -118,5 +120,95 @@ private boolean submitToStairway(String flightId, TrackedResource resource) {
+        ++completedFlights;
+      }
+    }
+    return completedFlights;
+  }
+
+  /**
+   * Completes the flight if Stairway has finished the flight non-fatally. Returns whether we were
+   * able to update the state of the resource and flight successfully for the completed flight.
+   */
+  private boolean completeFlight(JanitorDao.TrackedResourceAndFlight resourceAndFlight) {
+    String flightId = resourceAndFlight.cleanupFlight().flightId();
+    FlightState flightState;
+    try {
+      flightState = stairway.getFlightState(flightId);
+    } catch (DatabaseOperationException | InterruptedException e) {
+      logger.error(String.format("Error getting state of finishing flight [%s]", flightId), e);
+      return false;
+    }
+    TrackedResourceState endCleaningState;
+    if (flightState.getFlightStatus().equals(FlightStatus.SUCCESS)) {
+      endCleaningState = TrackedResourceState.DONE;
+    } else if (flightState.getFlightStatus().equals(FlightStatus.ERROR)) {
+      endCleaningState = TrackedResourceState.ERROR;
+    } else {
+      // The flight hasn't finished or has finished fatally. Let Stairway keep working or the fatal
+      // monitor handle this respectively.
+      return false;
+    }
+    try {
+      updateFinishedCleanupState(
+          resourceAndFlight.trackedResource().trackedResourceId(), flightId, endCleaningState);
+    } catch (UnexpectedCleanupState unexpectedCleanupState) {
+      logger.error(
+          String.format("Error finishing flight changing cleanup state. [%s]", flightId),
+          unexpectedCleanupState);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Transactionally update the TrackedResourceState and CleanupFlightState for the finishing
+   * flight. Throws an exception if there is an unexpected state to rollback the transaction.
+   */
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  private void updateFinishedCleanupState(
+      TrackedResourceId trackedResourceId, String flightId, TrackedResourceState endCleaningState)
+      throws UnexpectedCleanupState {
+    // Retrieve the TrackedResource within this transaction to ensure no one else is abandoning or
+    // duplicating it while we finish it.
+    Optional<TrackedResource> resource = janitorDao.retrieveTrackedResource(trackedResourceId);
+    if (!resource.isPresent()) {
+      throw new UnexpectedCleanupState(
+          String.format("Unable to find tracked_resource with id [%s]", trackedResourceId));
+    }
+    TrackedResourceState resourceState = resource.get().trackedResourceState();
+    if (resourceState.equals(TrackedResourceState.CLEANING)) {
+      janitorDao.updateResourceState(trackedResourceId, endCleaningState);
+      // We assume no one else is modifying the CleanupFlightState while we do this.
+    } else if (!resourceState.equals(TrackedResourceState.ABANDONED)
+        && !resourceState.equals(TrackedResourceState.DUPLICATED)) {
```

Comment:
Having duplicates state is expected. Feel this should not be an exception.

Reply:
Not sure what you mean, there's a `!` negation. Anything that's *not* ABANDONED and DUPLICATED is not expected. DUPLICATED is expected and does not throw.